### PR TITLE
[DataLoader] Adding ability to use dill to pass DataPipes in mutiprocessing

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #77288

Fixes part of https://github.com/pytorch/data/issues/397 (only DataLoader v2 remaining)

Differential Revision: [D36347737](https://our.internmc.facebook.com/intern/diff/D36347737)

### CODE REVIEW

File: torch/utils/data/datapipes/datapipe.py [hunk 1]
```diff
@@ -265,3 +278,43 @@ def __iter__(self) -> Iterator[T]:
     def raw_iterator(self) -> T:  # type: ignore[misc]
         for i in self.items:
             yield i
+
+class TravelingDataPipe:
```
Comment (ejguan) [inline torch/utils/data/datapipes/datapipe.py:282]: Could we rename it to something like `_DataPipeSerializationWrapper`? With `DataPipe` at the end of this class, we might give users a wrong hint. And, make it private (underscore) might be even better.

**Review by ejguan — APPROVED** [review]
LGTM with a couple comments below!

File: torch/utils/data/datapipes/datapipe.py (hunk 1 above)
Comment (NivekT) [reply torch/utils/data/datapipes/datapipe.py:282]: I agree with the renaming, we might want to rename `TravelingIterDataPipe` and `TravelingMapDataPipe` too, especially if these names show up in the DataPipe graph and users want to know what they are

**Review by NivekT — APPROVED** [review]

File: test/test_datapipe.py [hunk 2]
```diff
@@ -2082,6 +2082,23 @@ def test_traverse_mixdatapipe(self):
         expected: Dict[Any, Any] = {iter_dp: {source_map_dp: {}}}
         self.assertEqual(expected, graph)
 
+def unbatch(x):
+    return x[0]
+
+class TestSerialization(TestCase):
+    @skipIfNoDill
```
[the remark this answers is not in the document]
Comment (ejguan) [reply test/test_datapipe.py:2091]: Lol. I completely mess up the `spawn` and `fork`. Sorry abotu that

**Comment by NivekT** [comment]
I think users can also use the same pattern here if they ever want to have custom serialization logic? I am anticipating that they can follow this pattern, implement a DataPipe with custom `__getstate__` and `__setstate__`, put that at the end of the pipeline, and the rest should work.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36347737).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36347737).

File: torch/utils/data/communication/eventloop.py [hunk 3]
```diff
@@ -44,8 +55,15 @@ def SpawnThreadForDataPipeline(datapipe):
...
     try:
         new_datapipe = pickle.loads(pickle.dumps(datapipe))
-    except Exception as e:
-        raise Exception('Unable to pickle DataPipe to make thread local copy', e)
+    except Exception as pe:
+        if HAS_DILL:
+            try:
+                new_datapipe = dill.loads(dill.dumps(datapipe))
+            except Exception as de:
+                Exception('Unable to dill DataPipe to make thread local copy', de)
```
[the remark this answers is not in the document]
Comment (VitalyFedyunin) [reply torch/utils/data/communication/eventloop.py:63]: Good catch, I will follow-up with quick patch.

**Pull request closed without merging**